### PR TITLE
FREEZE TABLE syntax returning snapshot details

### LINKED ISSUE #7785: ALTER TABLE FREEZE should return snapshot ID as result
Running `ALTER TABLE FREEZE` currently does not return the incremental snapshot ID. I can only get it from the `./shadow/increment.txt` file:

```
localhost :) ALTER TABLE stats FREEZE

ALTER TABLE stats
    FREEZE


Ok.

0 rows in set. Elapsed: 0.018 sec.
```
Assuming there are multiple processes making snapshots, I would like to know the snapshot ID when issuing the FREEZE query. Otherwise I can not be sure, which snapshot was the one that I issued.

What I would like to see:
```
localhost :) ALTER TABLE stats FREEZE

ALTER TABLE stats
    FREEZE

┌─snapshot_id─┐
│           3 │
└─────────────┘

1 rows in set. Elapsed: 0.001 sec.
```

Alternatively, also the 'outcome' could be reported (ok, error, ...) in the result.

### PR DESCRIPTION
**Pull request by nvartolomei**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (up to few sentences, required except for Non-significant/Documentation categories):

`FREEZE TABLE [db.]name [PARTITION partition_expr] [WITH NAME custom_dir]`

Detailed description (optional):

```sql
FREEZE TABLE system.query_log

┌─backup_name─┬─partition─┬─partition_id─┬─part_name──────┬─backup_path─────────────────────────────────────────────────────────────────────────┐
│ 5           │ 201912    │ 201912       │ 201912_1_10_2  │ /home/nv/clickhouse-testbed/srv2/data/shadow/5/data/system/query_log/201912_1_10_2  │
│ 5           │ 201912    │ 201912       │ 201912_11_11_0 │ /home/nv/clickhouse-testbed/srv2/data/shadow/5/data/system/query_log/201912_11_11_0 │
└─────────────┴───────────┴──────────────┴────────────────┴─────────────────────────────────────────────────────────────────────────────────────┘

2 rows in set. Elapsed: 0.007 sec.
```

Closes #7785.

### CODE REVIEW

**Comment by nvartolomei** [comment]
CI is hibernating?

**Comment by nvartolomei** [comment]
I'll avoid rebasing this until it gets a green light, did it too many times already.

File: dbms/Storages/MergeTree/MergeTreeData.h [hunk 1]
```diff
@@ -888,9 +884,10 @@ class MergeTreeData : public IStorage
     /// Checks whether the column is in the primary key, possibly wrapped in a chain of functions with single argument.
     bool isPrimaryOrMinMaxKeyColumnPossiblyWrappedInFunctions(const ASTPtr & node) const;
 
-    /// Common part for |freezePartition()| and |freezeAll()|.
+    /// Common part for |freeze()|.
     using MatcherFn = std::function<bool(const DataPartPtr &)>;
-    void freezePartitionsByMatcher(MatcherFn matcher, const String & with_name, const Context & context);
+    MatcherFn freezePartitionMatcher(const ASTPtr & partition_ast, const Context & context);
```
Comment (alesapin) [inline dbms/Storages/MergeTree/MergeTreeData.h:889]: `getFreezePartitionMatcher`?

File: dbms/Parsers/ParserFreezeQuery.cpp [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+
+namespace DB
+{
+bool ParserFreezeQuery::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
+{
+    ParserKeyword s_freeze_table("FREEZE TABLE");
+    ParserKeyword s_partition("PARTITION");
+
+    ParserKeyword s_with("WITH");
```
Comment (alesapin) [inline dbms/Parsers/ParserFreezeQuery.cpp:17]: Why not just one `"WITH NAME"` keyword?

File: dbms/Interpreters/InterpreterFreezeQuery.h [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+#include <Interpreters/IInterpreter.h>
+#include <Parsers/IAST_fwd.h>
+
+
+namespace DB
+{
+class Context;
+
+class InterpreterFreezeQuery : public IInterpreter
```
Comment (alesapin) [inline dbms/Interpreters/InterpreterFreezeQuery.h:11]: Comments.

File: dbms/Interpreters/InterpreterFreezeQuery.h [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+public:
+    InterpreterFreezeQuery(const ASTPtr & query_ptr_, Context & context_) : query_ptr(query_ptr_), context(context_) {}
+
+    BlockIO execute() override;
+
+    static Block getSampleBlock();
+
+private:
+    ASTPtr query_ptr;
+    Context & context;
```
Comment (alesapin) [inline dbms/Interpreters/InterpreterFreezeQuery.h:22]: `const Context &`?

File: dbms/Interpreters/InterpreterFreezeQuery.h [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+    BlockInputStreamPtr executeImpl();
```
Comment (alesapin) [inline dbms/Interpreters/InterpreterFreezeQuery.h:24]: Also const?

File: dbms/Storages/FreezeResult.h [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+#pragma once
+
+namespace DB
+{
+/// Result of FREEZE TABLE ... query.
+struct FreezeResult
+{
+    struct PartInfo
+    {
+        PartInfo(String partition_, String partition_id_, String part_name_, String backup_path_)
```
Comment (alesapin) [inline dbms/Storages/FreezeResult.h:12]: Why default `PartInfo{partition, partition_id, ...}` is not ok?

File: dbms/Interpreters/InterpreterFreezeQuery.cpp [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+BlockInputStreamPtr InterpreterFreezeQuery::executeImpl()
+{
+    const auto & ast = query_ptr->as<ASTFreezeQuery &>();
+
+    context.checkAccess(AccessType::FREEZE_PARTITION, ast.database, ast.table);
+
+    auto table_id = context.resolveStorageID(ast, Context::ResolveOrdinary);
+    StoragePtr table = DatabaseCatalog::instance().getTable(table_id);
+
+    FreezeResult freeze_result = table->freeze(ast.partition, ast.with_name, context);
```
Comment (alesapin) [inline dbms/Interpreters/InterpreterFreezeQuery.cpp:46]: maybe check access `w` to `args.with_name`?

File: dbms/Parsers/ParserFreezeQuery.cpp [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    if (s_with.ignore(pos, expected))
+    {
+        if (!s_name.ignore(pos, expected))
+            return false;
+
+        ASTPtr ast_with_name;
+        if (!string_literal_p.parse(pos, ast_with_name, expected))
+            return false;
+
+        query->with_name = ast_with_name->as<ASTLiteral &>().value.get<const String &>();
```
Comment (alesapin) [inline dbms/Parsers/ParserFreezeQuery.cpp:48]: `sageGet` is better in this case. Also, `with_name` will be correctly escaped and maybe https://github.com/ClickHouse/ClickHouse/pull/8076/files#diff-e463b478463f2d31fc254257f4e69376R23 is not required?

**Review by alesapin — COMMENTED** [review]
I'm not sure about the new query type, maybe we can extent `ALTER TABLE ... FREZEE PARTITION`? We can introduce setting `freeze_partition_verbose_output` if we don't want to break compatibility and restrict to use `FREEZE PARTITION` in a chain with other types of ALTER.
Otherwise, we need to deprecate `ALTER TABLE FREEZE PARTITION`. Maybe warning log message in logs as the first step?

**Comment by alexey-milovidov** [comment]
We decided to go this way: https://github.com/ClickHouse/ClickHouse/pull/8076#pullrequestreview-389025609

**Pull request closed without merging**